### PR TITLE
Update to API/Routes Design Pattern, Adding Login/Signup API, Tank API **Be sure to title your PRs next time :)**

### PR DESCRIPTION
**Pull request by Airor4**
**Description**
Added proxy, and changed the api/routes to a route/controller design pattern. User login and signup are ready to go. See the route files on how to call them. When testing the API, make sure you are running the backend server along with the frontend server. It may seem obvious but we can be forgetful hahah.

Fixes #(issue)
#43 #44 

**Type of change**
Check options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Fix or feature that would cause existing functionality to not work as expected
- [ ] This change requires a update in the design document

**Checklist:**
Go down the list and check them off.

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the design document (if applicable)
- [X] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/bug_report.md [hunk 1]
```diff
@@ -1,27 +0,0 @@
----
-name: Bug report
```
Comment (SecondThread) [inline .github/ISSUE_TEMPLATE/bug_report.md]: I'm pretty sure we don't want this deleted. @SleepyLime to confirm though

File: .github/ISSUE_TEMPLATE/feature_request.md [hunk 2]
```diff
@@ -1,17 +0,0 @@
----
-name: Feature request
```
Comment (SecondThread) [inline .github/ISSUE_TEMPLATE/feature_request.md]: Same thing; pretty sure we want this to still be here. @SleepyLime

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -1,22 +0,0 @@
-**Description**
-Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.
```
Comment (SecondThread) [inline .github/pull_request_template.md]: Ditto

File: src/backend/controllers/tankController.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+import type {
+    $Request,
+    $Response,
+    NextFunction,
+    Middleware,
+  } from 'express';
+
+exports.getFavorite = (req: $Request, res: $Response) => {
+    user.findById(req.params.userId, 'favoriteTankId', function(err, myUser){
```
Comment (SecondThread) [inline src/backend/controllers/tankController.js]: What are the types of err and myUser? Can you flow-type them please?

File: src/backend/controllers/userController.js [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        return res
+            .status(400)
+            .json({ errors: errors.array() });
+    }
+
+    // Assigns request body to user schema fields
+    const { userName, password } = req.body;
+
+    try{
+        // See if User exists - might change this to const
```
Comment (SecondThread) [inline src/backend/controllers/userController.js:98]: Yeah, change this to const please :)

File: src/backend/controllers/userController.js [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        try{
+            const user = await User.findById(req.params.userId);
+            res.json(user);
+        } catch(err){
+            console.error(err.message);
+            res.status(500).json({ msg: 'Server Error'});
+        }
+}
+
+/*exports.updateCassus = async (req: $Request, res: $Response){
```
Comment (SecondThread) [inline src/backend/controllers/userController.js]: Delete this if it isn't working and then publish it in a later push if you need to

File: src/backend/controllers/userController.js [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+    }
+    catch(err){
+
+    }
+}*/
+
+exports.checkAuth = async (req: $Request, res: $Response) => {
+    try{
+        const user = await User.findById(req.userId).select('-password');
```
Comment (SecondThread) [inline src/backend/controllers/userController.js:150]: Just curious: what is the hyphen for before password?

File: src/backend/routes/userRoutes.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+const auth = require('../middleware/auth');
+
+const userController = require('../controllers/userController');
+
+// Create a new user
+// Route call: registerUser
+// Req must include userName and password in body
+// Returns a jwt upon success
+router.post('/registerUser', [
+    check('userName', 'Please enter a username with 3 or more characters')
```
Comment (SecondThread) [inline src/backend/routes/userRoutes.js:21]: Clean, I like it

**Review by SecondThread — CHANGES_REQUESTED** [review]
I left comments for some of the things I think you should change.

Also, did you test this stuff to make sure it works? Can you write a doc somewhere so we can see what we need to do to verify that your stuff works?

File: src/backend/controllers/userController.js (hunk 7 above)
Comment (Airor4) [reply src/backend/controllers/userController.js:150]: It excludes the password document field.

File: .github/ISSUE_TEMPLATE/bug_report.md (hunk 1 above)
Comment (Airor4) [reply .github/ISSUE_TEMPLATE/bug_report.md]: I'm not sure how this was deleted or how to get it back.

File: src/backend/controllers/userController.js (hunk 7 above)
Comment (SecondThread) [reply src/backend/controllers/userController.js:150]: oh lol that's probably important

**Comment by Derplime** [comment]
@Airor4 

So for recovering the `.github` file, you could see if it exists in another branch (i.e. `master`), copy it locally to your computer, go back to your branch, paste that in, and push that up to your remote. Assuming you didn't change any files in that `.github` file, that should remove that change compared to master.

File: src/.env [hunk 9]
```diff
@@ -0,0 +1,2 @@
+SECRET=supahsecret4
+PORT=3001
```
Comment (Derplime) [inline src/.env]: Request new line at end of file.

File: src/login/Login.css [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+.signuppopup {
+    padding: 5px 50px 20px 50px;
+    background: #004e92;
+}
+
+.infopopup {
+    padding: 5px 50px 20px 50px;
+    background: #004e92;
+}
```
Comment (Derplime) [inline src/login/Login.css]: Another new line here please. Sorry for pickiness.

**Review by Derplime — CHANGES_REQUESTED** [review]
So outside the changes @SecondThread requested, and my stupid picky ones, I'd also second on having a set of instructions to test this correctly before giving my approval.

**Review by SecondThread — APPROVED** [review]
lgtm

**Review by Derplime — APPROVED** [review]
👍  Forgot to put mine in as well.

**Comment by Derplime** [comment]
Merging this myself since I'll need this to continue my work.

**Pull request merged**